Translation of blog post: "2019-08-08-react-v16.9.0" (#317)
reactjs/pt-BR.reactjs.org: content/blog/2019-08-08-react-v16.9.0.md

---
title: "React v16.9.0 e a atualização do Roadmap"
author: [gaearon, bvaughn]
---

> Hoje estamos lançando o React 16.9. Ele contém vários novos recursos, correções de bugs e novos avisos de descontinuação para ajudar a se preparar para uma futura versão major.

Reviewer: ```suggestion
Hoje estamos lançando o React 16.9. Ele contém vários novos recursos, correções de bugs e novos avisos de depreciação para ajudar a se preparar para uma futura versão principal.
```

> ## Novas Descontinuações {#new-deprecations}

Reviewer: ```suggestion
## Novas Depreciações {#new-deprecations}
```

> ### Renomeado Métodos de Ciclo de Vida Inseguros {#renaming-unsafe-lifecycle-methods}

Reviewer: ```suggestion
### Renomeando Métodos de Ciclo de Vida Inseguros {#renaming-unsafe-lifecycle-methods}
```

> [Há mais de um ano](/blog/2018/03/27/update-on-async-rendering.html), anunciamos os métodos de ciclo de vida inseguros que estão sendo renomeados:

Reviewer: ```suggestion
[Há mais de um ano](/blog/2018/03/27/update-on-async-rendering.html), anunciamos que os métodos de ciclo de vida inseguros estão sendo renomeados:
```

* `componentWillMount` → `UNSAFE_componentWillMount`
* `componentWillReceiveProps` → `UNSAFE_componentWillReceiveProps`
* `componentWillUpdate` → `UNSAFE_componentWillUpdate`

**O React 16.9 não contém alterações significativas e os nomes antigos continuam funcionando nesta versão.** Mas agora você verá um aviso ao usar qualquer um dos nomes antigos:

![Warning: componentWillMount foi renomeado e não é recomendado para uso.](https://i.imgur.com/sngxSML.png)

Como o aviso sugere, geralmente há [abordagens melhores](/blog/2018/03/27/update-on-async-rendering.html#migrating-from-legacy-lifecycles) para cada um dos métodos inseguros. No entanto, talvez você não tenha tempo para migrar ou testar esses componentes. Nesse caso, recomendamos a execução de um script ["codemod"](https://medium.com/@cpojer/effective-javascript-codemods-5a6686bb46fb) que os renomeia automaticamente:

```bash
cd seu_projeto
npx react-codemod rename-unsafe-lifecycles
```

> *(Note queele diz `npx`, nãp `npm`. `npx` é um utilitário que vem com  Node 6+ por padrão.)*

Reviewer: ```suggestion
*(Note que ele diz `npx`, não `npm`. `npx` é um utilitário que vem com  Node 6+ por padrão.)*
```

> A execução deste codemod substituirá os nomes antigos `componentWillMount` pelos novos nomes como `UNSAFE_componentWillMount`:

Reviewer: ```suggestion
A execução deste codemod substituirá os nomes antigos como `componentWillMount` pelos novos nomes como `UNSAFE_componentWillMount`:
```

![Codemod em ação](https://i.imgur.com/Heyvcyi.gif)

> Os novos nomes como `UNSAFE_componentWillMount` **irão continuar a trabalhar em ambos React 16.9 e em React 17.x**. No entanto, o novo prefixo `UNSAFE_` ajudará os componentes com padrões problemáticos a se destacarem durante os code review e depuração de código. (Se desejar, você pode desestimular ainda mais o uso dele dentro do seu aplicativo com o [Strict Mode](/docs/strict-mode.html).)

Reviewer: ```suggestion
Os novos nomes como `UNSAFE_componentWillMount` **irão continuar a funcionar em ambos React 16.9 e em React 17.x**. No entanto, o novo prefixo `UNSAFE_` ajudará os componentes com padrões problemáticos a se destacarem durante os code review e depuração de código. (Se desejar, você pode desestimular ainda mais o uso dele dentro do seu aplicativo com o [Strict Mode](/docs/strict-mode.html).)
```

>Nota
>
>Saiba mais sobre nossa [política de versão e compromisso com a estabilidade](/docs/faq-versioning.html#commitment-to-stability).

> ### Depreciado `javascript:` URLs {#deprecating-javascript-urls}

Reviewer: ```suggestion
### Depreciando URLs `javascript:` {#deprecating-javascript-urls}
```

> As URLs que começam com `javascript:` são uma superfície de ataque perigosa porque é fácil incluir acidentalmente uma saída não-analisada em uma tag `<a href>` e criar uma brecha de segurança:

Reviewer: ```suggestion
As URLs que começam com `javascript:` são uma superfície de ataque perigosa porque é fácil incluir acidentalmente uma saída não-tratada em uma tag `<a href>` e criar uma brecha de segurança:
```

```js
const userProfile = {
  website: "javascript: alert('you got hacked')",
};
// This will now warn:
<a href={userProfile.website}>Profile</a>
```

> **Em React 16.9,** esse padrão continua a funcionar, mas registrará um aviso. Se você usar `javascript:` na lógica das URLs, tente usar manipuladores de eventos React. (Como último recurso, você pode contornar a proteção [`dangerouslySetInnerHTML`](/docs/dom-elements.html#dangerouslysetinnerhtml), mas é altamente desenvorajado e muitas vezes leva a falhas de segurança.)

Reviewer: ```suggestion
**Em React 16.9,** esse padrão continua a funcionar, mas registrará um aviso. Se você usar `javascript:` na lógica das URLs, tente usar manipuladores de eventos React. (Como último recurso, você pode contornar a proteção [`dangerouslySetInnerHTML`](/docs/dom-elements.html#dangerouslysetinnerhtml), mas é altamente desencorajado e muitas vezes leva a falhas de segurança.)
```

> **Em uma versão major futura,** React lançará um erro se encontrar um `javascript:` URL.

Reviewer: ```suggestion
**Em uma versão principal futura,** React lançará um erro se encontrar um `javascript:` URL.
```

> ### Depreciado "Factory" Componentes {#deprecating-factory-components}

Reviewer: ```suggestion
### Depreciando "Factory" Componentes {#deprecating-factory-components}
```

> Antes de compilar as classes JavaScript com o Babel tornou-se popular, o React tinha suporte para um componente "factory" que retorna um objeto com um `render` método:

Reviewer: ```suggestion
Antes de compilar as classes JavaScript com o Babel se tornar popular, o React tinha suporte para um componente "factory" que retorna um objeto com um método `render`:
```

```js
function FactoryComponent() {
  return { render() { return <div />; } }
}
```

Esse padrão é confuso porque parece muito com um componente de função - mas não é um. (Um componente de função apenas retornaria o `<div />` do exemplo acima.)

> Esse padrão quase nunca foi usado por ninguem, e o suporte faz com que o React seja um pouco maior e mais lento do que o necessário. Então estamos depreciando este padrão na 16.9 e registrando um aviso se ele for encontrado. Se você confiar nele, a adição `FactoryComponent.prototype = React.Component.prototype` pode servir como uma solução alternativa. Como alternativa, você pode convertê-lo em uma classe ou em um componente de função.

Reviewer: ```suggestion
Esse padrão quase nunca foi usado e o suportá-lo faz com que o React seja um pouco maior e mais lento do que o necessário. Então estamos depreciando este padrão na 16.9 e registrando um aviso se ele for encontrado. Se você confiar nele, a adição `FactoryComponent.prototype = React.Component.prototype` pode servir como uma solução alternativa. Alternativamente, você pode convertê-lo em uma classe ou em um componente de função.
```

> Não esperamos que a maioria das bases de código sejam afetadas por isso.

Reviewer: ```suggestion
Não esperamos que a maioria das bases de código seja afetadas por isso.
```

> 

Reviewer: Ficou faltando esta linha aqui, mas compensa mais remover na versão em Inglês.

Author: ahh sim, eu abri um PR para ajudar isso na versão em inglês >> reactjs/reactjs.org#2294

> ## Novas Features {#new-features}

Reviewer: Talvez seja melhor colocar funcionalidades aqui?

Author: Pode ser, eu acabei fazendo de madrugada a tradução, vi que muitas das correções que tu fez foi erros simples hehehe, provavel eu estava quase dormindo na hora, vou ajustar para funcionalidades:

```suggestion
## Novas Funcionalidades {#new-features}
```

### Assíncrono [`act()`](/docs/test-utils.html#act) para teste {#async-act-for-testing}

> [O React 16.8](/blog/2019/02/06/react-v16.8.0.html) introduzio um novo utilitário de teste chamado [`act()`](/docs/test-utils.html#act) para ajudá-lo a escrever testes que correspondam melhor ao comportamento do navegador. Por exemplo, várias atualizações de estado em um único `act()` lote de coletas. Isso corresponde ao modo como o React já funciona ao manipular eventos reais do navegador e ajuda a preparar seus componentes para o futuro, no qual o React será atualizado em lote com mais frequência.

Reviewer: ```suggestion
[O React 16.8](/blog/2019/02/06/react-v16.8.0.html) introduziu um novo utilitário de teste chamado [`act()`](/docs/test-utils.html#act) para ajudá-lo a escrever testes que correspondam melhor ao comportamento do navegador. Por exemplo, várias atualizações de estado em um único `act()` são enfileiradas. Isso corresponde ao modo como o React já funciona ao manipular eventos reais do navegador e ajuda a preparar seus componentes para o futuro, no qual o React irá enfileirar atualizações com mais frequência.
```

No entanto, em 16.8 `act()` apenas suporta funções síncronas. Às vezes, você pode ter visto um aviso como este em um teste, mas [não conseguiu corrigi-lo facilmente](https://github.com/facebook/react/issues/14769):

```

> Uma atualização para SomeComponente dentro de um teste não foi encapsulada no act(...).

Reviewer 2: Aqui é a mensagem de erro no console. Será que é necessário traduzir?!

Author: ```suggestion
An update to SomeComponent inside a test was not wrapped in act(...).
```

```

**No React 16.9, `act()` também aceita funções assíncronas,** e você pode chamar o `await`:

```js
await act(async () => {
  // ...
});
```

Isso resolve os casos restantes em que antes você não podia usar o `act()`, como quando a atualização do estado estava dentro de uma função assíncrona. Como resultado, **você poderá corrigir todos os avisos do `act()` restantes em seus testes agora.** 

> Ouvimos que não havia informações suficiesntes sobre como escrever testes com `act()`. O novo guia [Receitas de Testes](/docs/testing-recipes.html) descreve cenários comuns e como `act()` pode ajudar você a escrever bons testes. Esses exemplos usam APIs vanilla do DOM, mas você pode também usar [a bibliteca React Testing](https://testing-library.com/docs/react-testing-library/intro) para reduzir o boilerplate de código. Muitos de seus métodos já usam `act()` internamente.

Reviewer: ```suggestion
Ouvimos que não haviam informações suficientes sobre como escrever testes com `act()`. O novo guia [Receitas de Testes](/docs/testing-recipes.html) descreve cenários comuns e como `act()` pode ajudar você a escrever bons testes. Esses exemplos usam APIs vanilla do DOM, mas você pode também usar [a bibliteca React Testing](https://testing-library.com/docs/react-testing-library/intro) para reduzir o boilerplate de código. Muitos de seus métodos já usam `act()` internamente.
```

> Ouvimos que não haviam informações suficientes sobre como escrever testes com `act()`. O novo guia [Receitas de Testes](/docs/testing-recipes.html) descreve cenários comuns e como `act()` pode ajudar você a escrever bons testes. Esses exemplos usam APIs vanilla do DOM, mas você pode também usar [a bibliteca React Testing](https://testing-library.com/docs/react-testing-library/intro) para reduzir o boilerplate de código. Muitos de seus métodos já usam `act()` internamente.

Reviewer 2: `React Testing Library` é o nome da biblioteca. Acho que não é necessário traduzir.
```suggestion
Ouvimos que não haviam informações suficientes sobre como escrever testes com `act()`. O novo guia [Receitas de Testes](/docs/testing-recipes.html) descreve cenários comuns e como `act()` pode ajudar você a escrever bons testes. Esses exemplos usam APIs vanilla do DOM, mas você pode também usar o [React Testing Library](https://testing-library.com/docs/react-testing-library/intro) para reduzir o boilerplate de código. Muitos de seus métodos já usam `act()` internamente.
```

> Por favor, informe-nos [com uma issue](https://github.com/facebook/react/issues) se você se deparar com outros cenários em que `act()` não funciona bem para você e tentaremos ajuda-lo.

Reviewer: ```suggestion
Por favor, informe-nos [com uma issue](https://github.com/facebook/react/issues) se você se deparar com outros cenários em que `act()` não funciona bem para você e tentaremos ajudá-lo.
```

### Medições de Desempenho com [`<React.Profiler>`](/docs/profiler.html) {#performance-measurements-with-reactprofiler}

> Em React 16.5, introduzimos um novo [React Profiler para o DevTools](/blog/2018/09/10/introducing-the-react-profiler.html) que ajude a encontrar gargalos de desempenho em seu aplicativo. **Em React 16.9, também estamos adicionando uma maneira *programática* de reunir as medições** chamadas `<React.Profiler>`. Esperamos que a maioria dos aplicativos menores não os use, mas pode ser útil rastrear as regressões de desempenho ao longo do tempo em aplicativos maiores.

Reviewer: ```suggestion
Em React 16.5, introduzimos um novo [React Profiler para o DevTools](/blog/2018/09/10/introducing-the-react-profiler.html) que ajuda a encontrar gargalos de desempenho em seu aplicativo. **Em React 16.9, também estamos adicionando uma maneira *programática* de reunir as medições** chamadas `<React.Profiler>`. Esperamos que a maioria dos aplicativos menores não os use, mas pode ser útil rastrear as regressões de desempenho ao longo do tempo em aplicativos maiores.
```

> O `<Profiler>` medido quantas vezes é renderizado em um aplicativo React e qual o "custo" da renderização. Sua finalidade é ajudar identificar partes de um aplicativo que é lento e podem se beneficiar da [otimização, como a memorização](/docs/hooks-faq.html#how-to-memoize-calculations).

Reviewer: ```suggestion
O `<Profiler>` mede com qual frequência uma aplicação React renderiza e qual o "custo" da renderização. Sua finalidade é ajudar identificar partes de um aplicativo que é lento e podem se beneficiar da [otimização, como a memorização](/docs/hooks-faq.html#how-to-memoize-calculations).
```

> O `<Profiler>` mede com qual frequência uma aplicação React renderiza e qual o "custo" da renderização. Sua finalidade é ajudar identificar partes de um aplicativo que é lento e podem se beneficiar da [otimização, como a memorização](/docs/hooks-faq.html#how-to-memoize-calculations).

Reviewer 2: `lento` aqui se refere às partes do aplicativo.
```suggestion
O `<Profiler>` mede com qual frequência uma aplicação React renderiza e qual o "custo" da renderização. Sua finalidade é ajudar identificar partes de um aplicativo que estão lentas e podem se beneficiar da [otimização, como a memorização](/docs/hooks-faq.html#how-to-memoize-calculations).
```

> Um `<Profiler>` pode ser adicionado em qualquer lugar em uma árvore React para medir o custo de renderizar essa parte da árvore.

Reviewer 2: Sugestão:
```suggestion
Um `<Profiler>` pode ser adicionado em qualquer lugar em uma árvore React para medir o custo de renderização dessa parte da árvore.
```

> Ele requer dois suportes: um `id` (string) e um [`onRender` callback](/docs/profiler.html#onrender-callback) (função) que o React chama toda vez que um componente dentro da árvore "compromete" uma atualização.

Reviewer 2: `Props` precisa de tradução? Achei `suportes` muito estranho.
`compromete` não parece ser a melhor tradução para `commits` aqui. Acho que algo assim é melhor:
```suggestion
Ele requer duas props: um `id` (string) e um [`onRender` callback](/docs/profiler.html#onrender-callback) (função) que o React chama toda vez que um componente dentro da árvore "envia" uma atualização.
```

);
```

Para saber mais sobre o `Profiler` e os parametros passado para o `onRender` callback, confira [a documentação do `Profiler`](/docs/profiler.html).

> Nota:
>

> > Profiling adiciona uma sobrecarga extra, portanto, **ela é desativada no [build de produção](https://reactjs.org/docs/optimizing-performance.html#use-the-production-build)**.

Reviewer 2: Em outros locais está usando compilação, então acho que faz sentido aqui também.
```suggestion
> Profiling adiciona uma sobrecarga extra, portanto, **ela é desativada na [compilação de produção](https://reactjs.org/docs/optimizing-performance.html#use-the-production-build)**.
```

> > A criação de perfil adiciona alguma sobrecarga adicional, portanto, **ela é desativada no [build de produção](https://reactjs.org/docs/optimizing-performance.html#use-the-production-build)**.

Reviewer: Não sei se criação de perfil é a melhor tradução para profiling.
Talvez possamos deixar em inglês.

Author: o que acha?

```suggestion
> Profiling adiciona uma sobrecarga extra, portanto, **ela é desativada no [build de produção](https://reactjs.org/docs/optimizing-performance.html#use-the-production-build)**.
```

Reviewer: Bom!

>

> > Para optar pelo perfil de produção, o React fornece uma compilação de produção especial com a criação de perfil ativada.

Reviewer: Mesma coisa aqui.

Author: o que acha?

```suggestion
> Para optar pelo profiling em produção, o React fornece uma compilação de produção especial com a criação de profiling ativo.
```

Reviewer: Gostei! ficou bem mais claro.

> Leia mais sobre como usar essa versão em [fb.me/react-profiling](https://fb.me/react-profiling).

## Bugfixes Notáveis {#notable-bugfixes}

Esta versão contém algumas outras melhorias notáveis:

> * Um problema quando chamar `findDOMNode()` dentro de um `<Suspense>` [foi corrigido](https://github.com/facebook/react/pull/15312).

Reviewer 2: ```suggestion
* Um problema ao chamar `findDOMNode()` dentro de um `<Suspense>` [foi corrigido](https://github.com/facebook/react/pull/15312).
```

* Um vazamento de memória causado pela retenção de subárvores excluídas [também foi corrigido](https://github.com/facebook/react/pull/16115).

> * Um loop infinito causado por `setState` em `useEffect` agora [registra um error](https://github.com/facebook/react/pull/15180). (Isto é semelhante ao erro que você vê quando você chama `setState` em `componentDidUpdate` de uma class.)

Reviewer: Talvez traduzir error aqui?

Reviewer: Traduzir classe também.

Author: ```suggestion
* Um loop infinito causado por `setState` em `useEffect` agora [registra um erro](https://github.com/facebook/react/pull/15180). (Isto é semelhante ao erro que você vê quando você chama `setState` em `componentDidUpdate` de uma classe.)
```

Agradecemos a todos os colaboradores que ajudaram a identificar e corrigir esses e outros problemas. Você pode encontrar o changelog completo [abaixo](#changelog).

## Uma atualização para o Roadmap {#an-update-to-the-roadmap}

Em [Novembro de 2018](/blog/2018/11/27/react-16-roadmap.html), publicamos este roteiro para as versões 16.x:

> * Uma versão minor de 16.x com React Hooks (estimativa passada: Q1 2019)

Reviewer: Quero sugerir que traduzamos minor version como `versão menor`.
Podemos colocar isso no Glossário?

Author: Acho que seria legal colocar no Glossário sim, vou fazer um commit com o ajuste neste PR logo na sequência.

* Uma versão minor de 16.x com o Concurrent Mode (estimativa passada: Q2 2019)
* Uma versão minor de 16.x com Suspense para Data Fetching (estimativa anterior: meados de 2019)

> Essas estimativas eram otimistas demais e precisávamos ajustá-las.

Reviewer: ```suggestion
Essas estimativas eram otimistas demais e precisamos ajustá-las.
```

> **tldr:** Nós enviamos os Hooks a tempo, mas estamso reagrupando o Concurrent Mode e Suspense para Data Fetching em uma única versão que pretendemos lançar ainda este ano.

Reviewer: ```suggestion
**tldr:** Nós lançamos os Hooks a tempo, mas estamos reagrupando o Concurrent Mode e Suspense para Data Fetching em uma única versão que pretendemos lançar ainda este ano.
```

> Em fevereiro, nós [lançamos a 16.8 um versão estável](/blog/2019/02/06/react-v16.8.0.html) incluindo o React Hooks, com suporte ao React Native [um mês mais tarde](https://facebook.github.io/react-native/blog/2019/03/12/releasing-react-native-059). No entando, subestimamos o trabalho de acompanhamento desta versão, incluindo as regras de lint, ferramentas de desenvolvimento, exemplos e mais documentação. Isso mudou a linha do tempo em alguns meses.

Reviewer: ```suggestion
Em fevereiro, nós [lançamos a versão estável 16.8](/blog/2019/02/06/react-v16.8.0.html) incluindo o React Hooks, com suporte ao React Native [um mês mais tarde](https://facebook.github.io/react-native/blog/2019/03/12/releasing-react-native-059). No entanto, subestimamos o trabalho de acompanhamento desta versão, incluindo as regras de lint, ferramentas de desenvolvimento, exemplos e mais documentação. Isso mudou a linha do tempo em alguns meses.
```

> Agora que React Hooks foram lançados, o trabalho no Concurrent Mode e Suspense para Data Fetching está em pleno andamento. O [novo site do Facebook que está atualmente em desenvolvimento ativo](https://twitter.com/facebook/status/1123322299418124289) é construído sobre esses recursos. Testá-los com código real ajudou a descobrir e resolver muitos problemas antes que eles pudessem afetar os usuários de código aberto. Algumas dessas correções envolveram uma reformulação interna desses recursos, o que também fez com que a linha do tempo fosse perdido.

Reviewer 2: ```suggestion
Agora que React Hooks foram lançados, o trabalho no Concurrent Mode e Suspense para Data Fetching está em pleno andamento. O [novo site do Facebook que está atualmente em desenvolvimento ativo](https://twitter.com/facebook/status/1123322299418124289) é construído sobre esses recursos. Testá-los com código real ajudou a descobrir e resolver muitos problemas antes que eles pudessem afetar os usuários de código aberto. Algumas dessas correções envolveram uma reformulação interna desses recursos, o que também fez com que a linha do tempo fosse alterada.
```

Com esse novo entendimento, aqui está o que planejamos fazer a seguir.

### Um Lançamento em Vez de Dois {#one-release-instead-of-two}

Concurrent Mode e Suspense [fortalecem o novo site do Facebook](https://developers.facebook.com/videos/2019/building-the-new-facebookcom-with-react-graphql-and-relay/) que está em desenvolvimento ativo, por isso estamos confiantes de que estão perto de um estado estável tecnicamente. Agora também entendemos melhor as etapas concretas antes que elas estejam prontas para adoção de código aberto.

> Originalmente, pensamos em dividir o Concurrent Mode e Suspense para Data Fetching em dois lançamentos. Descobrimos que esse sequenciamento é confuso para explicar porque esses recursos são mais relacionados do que pensávamos inicialemnte. Portanto, planejamos lançar suporte para o Concurrent Mode e Suspense para Data Fetching em uma única versão.

Reviewer 2: Typo
```suggestion
Originalmente, pensamos em dividir o Concurrent Mode e Suspense para Data Fetching em dois lançamentos. Descobrimos que esse sequenciamento é confuso para explicar porque esses recursos são mais relacionados do que pensávamos inicialmente. Portanto, planejamos lançar suporte para o Concurrent Mode e Suspense para Data Fetching em uma única versão.
```

> Nós não queremos prometer demais a data de lançamento novamente. Dado que confiamos em ambos no código de produção, esperamos fornecer uma versão 16.x com suporte opcional para eles este ano.

Reviewer 2: ```suggestion
Nós não queremos comprometer a data de lançamento novamente. Dado que confiamos em ambos no código de produção, esperamos fornecer uma versão 16.x com suporte opcional para eles este ano.
```

### Uma Atualização na Busca de Dados {#an-update-on-data-fetching}

> Embora o React não tenha opinião sobre como você busca dados, a primeira versão do Suspense para Data Fetching provavelmente enfocará a integração com *bibliotecas de busca de dados opinativas*. Por exemplo, no Facebook, estamos usando as APIs do Relay que se integram com o Suspense. Vamos documentar como outras bibliotecas opinativas como a Apollo podem suportar uma integração semelhante.

Reviewer: ```suggestion
Embora o React não tenha opinião sobre como você busca dados, a primeira versão do Suspense para Data Fetching provavelmente focará a integração com *bibliotecas de busca de dados opinativas*. Por exemplo, no Facebook, estamos usando as APIs do Relay que se integram com o Suspense. Vamos documentar como outras bibliotecas opinativas como a Apollo podem suportar uma integração semelhante.
```

Na primeira versão, *não* pretendemos focar na solução ad-hoc "fire an HTTP request" que usamos nas demos anteriores (também conhecidas como "React Cache"). No entanto, esperamos que tanto nós como a comunidade React exploremos esse espaço nos meses após o lançamento inicial.

### Uma Atualização na Renderização do Servidor {#an-update-on-server-rendering}

Iniciamos o trabalho no [novo renderizador de servidor com capacidade para suspense](/blog/2018/11/27/react-16-roadmap.html#suspense-for-server-rendering), mas *não* esperamos que ele esteja pronto para a versão inicial do Concurrent Mode. Essa versão, no entanto, fornecerá uma solução temporária que permitirá que o renderizador do servidor existente emita HTML para fallbacks de Suspense imediatamente e, em seguida, renderize seu conteúdo real no cliente. Esta é a solução que estamos atualmente usando no Facebook até que o renderizador de streaming esteja pronto.

### Por que Demora Tanto Tempo? {#why-is-it-taking-so-long}

> Lançamos as peças individuais que levaram o Concurrent Mode quando elas se tornaram estáveis, incluindo a [nova API de context](/blog/2018/03/29/react-v-16-3.html), [carregamento lento com Suspense](/blog/2018/10/23/react-v-16-6.html) e [Hooks](/blog/2019/02/06/react-v16.8.0.html). Também estamos ansiosos para liberar as outras partes que faltam, mas [testá-las em escala](/docs/design-principles.html#dogfooding) é uma parte importante do processo. A resposta honesta é que isso só levou mais trabalho do que esperávamos quando começamos. Como sempre, agradecemos suas dúvidas e comentários no [Twitter](https://twitter.com/reactjs) e em nossas [issues](https://github.com/facebook/react/issues).

Reviewer 2: ```suggestion
Lançamos as peças individuais que levaram ao Concurrent Mode quando elas se tornaram estáveis, incluindo a [nova API de contexto](/blog/2018/03/29/react-v-16-3.html), [carregamento sob demanda com Suspense](/blog/2018/10/23/react-v-16-6.html) e [Hooks](/blog/2019/02/06/react-v16.8.0.html). Também estamos ansiosos para liberar as outras partes que faltam, mas [testá-las em escala](/docs/design-principles.html#dogfooding) é uma parte importante do processo. A resposta honesta é que isso só levou mais trabalho do que esperávamos quando começamos. Como sempre, agradecemos suas dúvidas e comentários no [Twitter](https://twitter.com/reactjs) e em nossas [issues](https://github.com/facebook/react/issues).
```

> Enviamos as peças individuais que levaram o Concurrent Mode quando elas se tornaram estáveis, incluindo a [nova API de context](/blog/2018/03/29/react-v-16-3.html), [carregamento lento com Suspense](/blog/2018/10/23/react-v-16-6.html) e [Hooks](/blog/2019/02/06/react-v16.8.0.html). Também estamos ansiosos para liberar as outras partes que faltam, mas [testá-las em escala](/docs/design-principles.html#dogfooding) é uma parte importante do processo. A resposta honesta é que isso só levou mais trabalho do que esperávamos quando começamos. Como sempre, agradecemos suas dúvidas e comentários no [Twitter](https://twitter.com/reactjs) e em nossas [issues](https://github.com/facebook/react/issues).

Reviewer: ```suggestion
Lançamos as peças individuais que levaram o Concurrent Mode quando elas se tornaram estáveis, incluindo a [nova API de context](/blog/2018/03/29/react-v-16-3.html), [carregamento lento com Suspense](/blog/2018/10/23/react-v-16-6.html) e [Hooks](/blog/2019/02/06/react-v16.8.0.html). Também estamos ansiosos para liberar as outras partes que faltam, mas [testá-las em escala](/docs/design-principles.html#dogfooding) é uma parte importante do processo. A resposta honesta é que isso só levou mais trabalho do que esperávamos quando começamos. Como sempre, agradecemos suas dúvidas e comentários no [Twitter](https://twitter.com/reactjs) e em nossas [issues](https://github.com/facebook/react/issues).
```

## Instalação {#installation}

### React {#react}

React v16.9.0 está disponível no registro npm.

Para instalar o React 16 com Yarn, execute:

```bash
yarn add react@^16.9.0 react-dom@^16.9.0
```

Para instalar o React 16 com npm, execute:

```bash
npm install --save react@^16.9.0 react-dom@^16.9.0
```

Nós também fornecemos compilações UMD de React via CDN:

```html
<script crossorigin src="https://unpkg.com/react@16/umd/react.production.min.js"></script>
<script crossorigin src="https://unpkg.com/react-dom@16/umd/react-dom.production.min.js"></script>
```

Consulte a documentação para [instruções detalhadas de instalação](/docs/installation.html).

> ## Changelog {#changelog}

Reviewer: Em geral, mensagens de commit são no indicativo presente do indicativo, por isso que sugeri as alterações abaixo.

### React {#react}

> * Adicionado `<React.Profiler>` API para reunir medições de desempenho programaticamente. ([@bvaughn](https://github.com/bvaughn) em [#15172](https://github.com/facebook/react/pull/15172))

Reviewer: ```suggestion
* Adiciona API do `<React.Profiler>` para reunir medições de desempenho programaticamente. ([@bvaughn](https://github.com/bvaughn) em [#15172](https://github.com/facebook/react/pull/15172))
```

> * Remover `unstable_ConcurrentMode` em favor de `unstable_createRoot`. ([@acdlite](https://github.com/acdlite) em [#15532](https://github.com/facebook/react/pull/15532))

Reviewer: ```suggestion
* Remove `unstable_ConcurrentMode` em favor de `unstable_createRoot`. ([@acdlite](https://github.com/acdlite) em [#15532](https://github.com/facebook/react/pull/15532))
```

### React DOM

> * Descontinuar nomes antigos dos métodos do ciclo de vida para os `UNSAFE_*`. ([@bvaughn](https://github.com/bvaughn) em [#15186](https://github.com/facebook/react/pull/15186) e [@threepointone](https://github.com/threepointone) em [#16103](https://github.com/facebook/react/pull/16103))

Reviewer: ```suggestion
* Deprecia nomes antigos dos métodos do ciclo de vida para os `UNSAFE_*`. ([@bvaughn](https://github.com/bvaughn) em [#15186](https://github.com/facebook/react/pull/15186) e [@threepointone](https://github.com/threepointone) em [#16103](https://github.com/facebook/react/pull/16103))
```

> * Descontinuar `javascript:` URLs como uma superfície de ataque comum. ([@sebmarkbage](https://github.com/sebmarkbage) em [#15047](https://github.com/facebook/react/pull/15047))

Reviewer: ```suggestion
* Deprecia URLs `javascript:` como uma superfície de ataque comum. ([@sebmarkbage](https://github.com/sebmarkbage) em [#15047](https://github.com/facebook/react/pull/15047))
```

> * Descontinuar componentes incomuns do "module pattern" (factory). ([@sebmarkbage](https://github.com/sebmarkbage) em [#15145](https://github.com/facebook/react/pull/15145))

Reviewer: ```suggestion
* Deprecia componentes incomuns do "module pattern" (factory). ([@sebmarkbage](https://github.com/sebmarkbage) em [#15145](https://github.com/facebook/react/pull/15145))
```

> * Adicionado suporte para o atributo `disablePictureInPicture` em `<video>`. ([@eek](https://github.com/eek) em [#15334](https://github.com/facebook/react/pull/15334))

Reviewer: ```suggestion
* Adiciona suporte para o atributo `disablePictureInPicture` em `<video>`. ([@eek](https://github.com/eek) em [#15334](https://github.com/facebook/react/pull/15334))
```

> * Adicionado suporte para o evento `onLoad` no `<embed>`. ([@cherniavskii](https://github.com/cherniavskii) em [#15614](https://github.com/facebook/react/pull/15614))

Reviewer: ```suggestion
* Adiciona suporte para o evento `onLoad` no `<embed>`. ([@cherniavskii](https://github.com/cherniavskii) em [#15614](https://github.com/facebook/react/pull/15614))
```

> * Adicionado suporte para editar o estado com `useState` no DevTools. ([@bvaughn](https://github.com/bvaughn) em [#14906](https://github.com/facebook/react/pull/14906))

Reviewer: ```suggestion
* Adiciona suporte para editar estados criados por `useState` no DevTools. ([@bvaughn](https://github.com/bvaughn) em [#14906](https://github.com/facebook/react/pull/14906))
```

> * Adicionado suporte para alterar o Suspense no DevTools. ([@gaearon](https://github.com/gaearon) em [#15232](https://github.com/facebook/react/pull/15232))

Reviewer: ```suggestion
* Adiciona suporte para alterar o Suspense no DevTools. ([@gaearon](https://github.com/gaearon) em [#15232](https://github.com/facebook/react/pull/15232))
```

> * Aviso quando `setState` é chamado no `useEffect`, criando um loop. ([@gaearon](https://github.com/gaearon) em [#15180](https://github.com/facebook/react/pull/15180))

Reviewer: ```suggestion
* Avisa quando `setState` é chamado no `useEffect`, criando um loop. ([@gaearon](https://github.com/gaearon) em [#15180](https://github.com/facebook/react/pull/15180))
```

> * Corrigido um vazamento de memória. ([@paulshen](https://github.com/paulshen) em [#16115](https://github.com/facebook/react/pull/16115))

Reviewer: ```suggestion
* Corrige um vazamento de memória. ([@paulshen](https://github.com/paulshen) em [#16115](https://github.com/facebook/react/pull/16115))
```

> * Corrigido uma falha dentro do `findDOMNode` dos componentes envolvidos por `<Suspense>`. ([@acdlite](https://github.com/acdlite) em [#15312](https://github.com/facebook/react/pull/15312))

Reviewer: ```suggestion
* Corrige uma falha dentro do `findDOMNode` dos componentes envolvidos por `<Suspense>`. ([@acdlite](https://github.com/acdlite) em [#15312](https://github.com/facebook/react/pull/15312))
```

> * Corrige efeitso pendentes de ser liberados muito tarde. ([@acdlite](https://github.com/acdlite) em [#15650](https://github.com/facebook/react/pull/15650))

Reviewer 2: Typo
```suggestion
* Corrige efeitos pendentes de ser liberados muito tarde. ([@acdlite](https://github.com/acdlite) em [#15650](https://github.com/facebook/react/pull/15650))
```

> * Corrigido efeitso pendentes de ser liberados muito tarde. ([@acdlite](https://github.com/acdlite) em [#15650](https://github.com/facebook/react/pull/15650))

Reviewer: ```suggestion
* Corrige efeitso pendentes de ser liberados muito tarde. ([@acdlite](https://github.com/acdlite) em [#15650](https://github.com/facebook/react/pull/15650))
```

> * Corrigido ordem de argumento incorreta em uma messagem de aviso. ([@brickspert](https://github.com/brickspert) em [#15345](https://github.com/facebook/react/pull/15345))

Reviewer: ```suggestion
* Corrige ordem de argumento incorreta em uma messagem de aviso. ([@brickspert](https://github.com/brickspert) em [#15345](https://github.com/facebook/react/pull/15345))
```

> * Corrigido a ocultação dos nós de fallback do Suspense quando houver um `!important` no estilo. ([@acdlite](https://github.com/acdlite) em [#15861](https://github.com/facebook/react/pull/15861) e [#15882](https://github.com/facebook/react/pull/15882))

Reviewer: ```suggestion
* Corrige a ocultação dos nós de fallback do Suspense quando houver um `!important` no estilo. ([@acdlite](https://github.com/acdlite) em [#15861](https://github.com/facebook/react/pull/15861) e [#15882](https://github.com/facebook/react/pull/15882))
```

> * Melhorado ligeiramente o desempenho de hidratação. ([@bmeurer](https://github.com/bmeurer) em [#15998](https://github.com/facebook/react/pull/15998))

Reviewer: ```suggestion
* Melhora ligeiramente o desempenho de hidratação. ([@bmeurer](https://github.com/bmeurer) em [#15998](https://github.com/facebook/react/pull/15998))
```

### React DOM Server

> * Corrigido a saída incorreta para nomes de propriedades CSS customizadas do camelCase. ([@bedakb](https://github.com/bedakb) em [#16167](https://github.com/facebook/react/pull/16167))

Reviewer: ```suggestion
* Corrige a saída incorreta para nomes de propriedades CSS customizadas do camelCase. ([@bedakb](https://github.com/bedakb) em [#16167](https://github.com/facebook/react/pull/16167))
```

### React Test Utilities e Test Renderer

> * Adicionado `act(async () => ...)` para testar atualizações de estado assíncronas. ([@threepointone](https://github.com/threepointone) em [#14853](https://github.com/facebook/react/pull/14853))

Reviewer: ```suggestion
* Adiciona`act(async () => ...)` para testar atualizações de estado assíncronas. ([@threepointone](https://github.com/threepointone) em [#14853](https://github.com/facebook/react/pull/14853))
```

> * Adicionado suporte para aninhamento `act` de diferentes renderizadores. ([@threepointone](https://github.com/threepointone) em [#16039](https://github.com/facebook/react/pull/16039) e [#16042](https://github.com/facebook/react/pull/16042))

Reviewer: ```suggestion
* Adiciona suporte para aninhamento `act` de diferentes renderizadores. ([@threepointone](https://github.com/threepointone) em [#16039](https://github.com/facebook/react/pull/16039) e [#16042](https://github.com/facebook/react/pull/16042))
```

> * Avisar no Strict Mode se os efeitso estiverem programados fora de uma chamada `act()`. ([@threepointone](https://github.com/threepointone) em [#15763](https://github.com/facebook/react/pull/15763) e [#16041](https://github.com/facebook/react/pull/16041))

Reviewer: ```suggestion
* Avisa no Strict Mode se os efeitos estiverem programados fora de uma chamada `act()`. ([@threepointone](https://github.com/threepointone) em [#15763](https://github.com/facebook/react/pull/15763) e [#16041](https://github.com/facebook/react/pull/16041))
```

> * Avisar ao usar `act` no renderizador errado. ([@threepointone](https://github.com/threepointone) em [#15756](https://github.com/facebook/react/pull/15756))

Reviewer: ```suggestion
* Avisa ao usar `act` no renderizador errado. ([@threepointone](https://github.com/threepointone) em [#15756](https://github.com/facebook/react/pull/15756))
```
